### DISCUSSION TITLE
Why doesn't my robot move?

### DISCUSSION THREAD

**Discussion by TEDPOWERTRAIN** [pybricks/support#1670] [Q&A]
I have found this algorithm from a leet code user and I am trying to incorporate it with my ev3. The program completes successfully but the robot doesn't move. 
Any clues? 

```python 
#!/usr/bin/env pybricks-micropython
from pybricks.hubs import EV3Brick
from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor,
                                 InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile

# This program requires LEGO EV3 MicroPython v2.0 or higher.
# Click "Open user guide" on the EV3 extension tab for more information.

# Create your objects here.
ev3 = EV3Brick()

# Initialize the motors.
left_motor = Motor(Port.A)
right_motor = Motor(Port.D)
ultra_motor = Motor(Port.C)

# Initialize the drive base.
robot = DriveBase(left_motor, right_motor, wheel_diameter=55.5, axle_track=106)

# Initialize the sensors.
color_sensor = ColorSensor(Port.S3)
front_touch = TouchSensor(Port.S2)
ultra_sensor = UltrasonicSensor(Port.S1)
gyro_sensor = GyroSensor(Port.S4)

# Functions movement
def robot_move():
    if ultra_sensor.distance() > 200:
        robot.straight(295)
        return True
    else:
        return False

def turn_left():
    robot.turn(-90)

def turn_right():
    robot.turn(90)

class Solutions:
    def aroundMaze(self, robot):
        """
        :type robot: Robot
        :rtype: None 
        """
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        visited = set()

        def go_back():
            turn_right()
            turn_right()
            robot_move()
            turn_right()
            turn_right()

        def backtrack(x, y, direction):
            visited.add((x, y))

            for i in range(4):
                new_direction = (direction + i) % 4

                new_x = x + directions[new_direction][0]
                new_y = y + directions[new_direction][1]

                if (new_x, new_y) not in visited and robot_move():
                    backtrack(new_x, new_y, new_direction)
                    go_back()

                turn_right()

        backtrack(0, 0, 0)

**Comment by laurensvalk** [comment]
> I have found this algorithm

What is the expected behavior of the program?

> The program completes successfully but the robot doesn't move.

This program defines a few functions and classes, but nothing that calls them. So nothing happens.

Maybe part of the code wasn't included?

**Reply by TEDPOWERTRAIN** [reply]
it is based on this https://leetcode.ca/2017-04-02-489-Robot-Room-Cleaner/

the robot is supposed to go around a room and visit every cell

**Reply by laurensvalk** [reply]
To get this code to work, have a look at the basics of Python functions and classes, and then have a look at what might be missing from your code. That part of it isn't specific to Pybricks, but the same as normal Python. This means you'll be able to find a lot of resources.

When you define a function, for example, you have to call it:

```python
# This defines a function but does not run it.
def hello(who):
    print("Hello:", who)

# Without this call, the program will do nothing.
hello("Ted")
```

**Reply by TEDPOWERTRAIN** [reply]
I just learnt python so it is a kind of mess. I will have a try with my robot once I am back to my pc. Thanks for helping me out

**Reply by TEDPOWERTRAIN** [reply]
From your perspective does the algorithm work?

**Reply by BertLindeman** [reply]
> From your perspective does the algorithm work?

Ted,

Like Laurens posted above, in my words: The code you show only creates building blocks of which a program can be built.
These are only functions that need to be called to create a real program.
So in short: without any additions this will not work.
But it **could** be.

Success,

Bert
